src/psyclone/psyir/backend/fortran.py

```diff
@@ -909,14 +1028,19 @@ def routine_node(self, node):
         # point we have to unify all declarations and resolve possible name
         # clashes that appear when merging the scopes.
         whole_routine_scope = SymbolTable(node)
+
         for schedule in node.walk(Schedule):
-            for symbol in schedule.symbol_table.symbols:
+            for symbol in schedule.symbol_table.symbols[:]:
                 try:
                     whole_routine_scope.add(symbol)
                 except KeyError:
-                    schedule.symbol_table.rename_symbol(
-                        symbol,
-                        whole_routine_scope.next_available_name(symbol.name))
+                    new_name = whole_routine_scope.next_available_name(
+                        symbol.name)
+                    # Ensure that the new name isn't already in the current
+                    # symbol table.
+                    new_name = schedule.symbol_table.next_available_name(
```

Comment:
But now could new_name be in whole_routine_scope? If a solution is not easy maybe we could just break at this point.

Also the coverage does not complain but could you add a test where both next_available_name have to change the name? (unless I missed it)

Reply:
Oh yes. I think I've got it now and have added a test.